File: vn/src/main/java/org/sqlbuilder/vn/objects/Predicate.java
package org.sqlbuilder.vn.objects;

import org.sqlbuilder.common.*;

import java.util.Comparator;
import java.util.Objects;

public class Predicate implements HasId, Insertable, Comparable<Predicate>
{
	public static final Comparator<Predicate> COMPARATOR = Comparator.comparing(Predicate::getName);

	public static final SetCollector2<Predicate> COLLECTOR = new SetCollector2<>(COMPARATOR);

	private final String name;

	// C O N S T R U C T O R

	public static Predicate make(final String name)
	{
		var p = new Predicate(name);
		COLLECTOR.add(p);
		return p;
	}

	private Predicate(final String name)
	{
		this.name = name;
	}

	// A C C E S S

	public String getName()
	{
		return name;
	}

	@Override
	public Integer getIntId()
	{
		return COLLECTOR.apply(this);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Predicate that = (Predicate) o;
		return name.equals(that.name);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(name);
	}

	// O R D E R I N G

	@Override
	public int compareTo(@NotNull final Predicate that)
	{
		return COMPARATOR.compare(this, that);
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		// id
		// name
		return String.format("'%s'", name);
	}


}

File: vn/src/main/java/org/sqlbuilder/vn/objects/Predicate.kt
package org.sqlbuilder.vn.objects

import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import java.util.*

class Predicate private constructor(
	@JvmField val name: String
) : HasId, Insertable, Comparable<Predicate> {

    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Predicate
        return name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }

    // O R D E R I N G

    override fun compareTo(that: Predicate): Int {
        return COMPARATOR.compare(this, that)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'$name'"
    }

    companion object {

        val COMPARATOR: Comparator<Predicate> = Comparator.comparing<Predicate, String> { it.name }

        @JvmField
        val COLLECTOR: SetCollector<Predicate> = SetCollector<Predicate>(COMPARATOR)

         fun make(name: String): Predicate {
            val p = Predicate(name)
            COLLECTOR.add(p)
            return p
        }
    }
}
